### PR TITLE
Binance methods

### LINKED ISSUE #5: Binance Bookkeeping methods

### PR DESCRIPTION
**Pull request by eanea**
closes #5

### CODE REVIEW

**Comment by catena2w** [comment]
close #5

**Review by catena2w — CHANGES_REQUESTED** [review]
Please extract common test code to a separate package. See https://github.com/Circlegain/exmods/issues/7

File: binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala [hunk 1]
```diff
@@ -24,14 +25,25 @@ final class Binance[F[_]: Monad: Clock](client: BinanceClient[F]) extends Exchan
...
-  def tradesHistory(pair: Option[TokenPair], paging: Paging): F[List[Trade]] = ???
+  def tradesHistory(pair: Option[TokenPair], paging: Paging): F[List[Trade]] = for {
+    params        <- BinanceTradeHistoryParams[F](pair, paging)
+    tradesHistory <- client.tradesHistory(params).flatMap(_.traverse(_.make[F, Trade]))
+  } yield tradesHistory
 
-  def transactionsHistory(token: Option[domain.Token], paging: Paging): F[List[Transaction]] = ???
+  def transactionsHistory(token: Option[domain.Token], paging: Paging): F[List[Transaction]] =
+    for {
+      depositHistory <- client
```
Comment (catena2w) [inline binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala:35]: As far as we anyway make 2 independent requests here, let's implement 2 methods `depositHistory` and `withdrawHistory` and combine them together in this interface method

File: binance/src/test/scala/BinanceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  it should "return balances" in {
+    val token     = "BTC"
+    val total     = BigDecimal(100)
+    val available = BigDecimal(100)
+    val balances0 = List(new Balance(new Currency(token), total.bigDecimal, available.bigDecimal))
+    val balances  = List(domain.Balance(ts, Token(token), available, total))
+    val client    = new FakeBinanceClient[IO](balances0, Nil, Nil, Nil)
+    println("client")
+    val probit    = new Binance[IO](client)
+    println("probit")
```
Comment (catena2w) [inline binance/src/test/scala/BinanceSpec.scala:22]: Pleases remove println and add rule to forbid it. 
Also  - why is it called probit here?

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    for {
+      curPair <- pairOp.map(_.make[F, CurrencyPair]).sequence
+      params = curPair
+                 .map(new BinanceTradeHistoryParams(_))
+                 .map { historyParams =>
+                   historyParams.setStartTime(paging.startTime.map(l => new Date(l)).orNull)
+                   historyParams
+                 }
+                 .map { historyParams =>
+                   historyParams.setEndTime(paging.endTime.map(l => new Date(l)).orNull)
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala:27]: Effectful statements shouldn't appear inside map

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import com.circlegain.exmods.binance.internal.instances._
+import com.circlegain.exmods.core.make.syntax._
+import com.circlegain.exmods.core.models.Paging
+import tofu.syntax.monadic._
+
+object BinanceTradeHistoryParams {
+
+  def apply[F[_]: Monad](pairOp: Option[TokenPair], paging: Paging): F[BinanceTradeHistoryParams] =
+    for {
+      curPair <- pairOp.map(_.make[F, CurrencyPair]).sequence
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala:19]:
```suggestion
      curPair <- pairOp.traverse(_.make[F, CurrencyPair])
```

File: binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala (hunk 1 above)
Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala]: done

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala (hunk 3 above)
Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala]: done

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala (hunk 4 above)
Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala]: :+1:

File: binance/src/test/scala/BinanceSpec.scala (hunk 2 above)
Comment (eanea) [reply binance/src/test/scala/BinanceSpec.scala]: dropped

**Review by catena2w — APPROVED** [review]

File: binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala [hunk 5]
```diff
@@ -24,14 +25,31 @@ final class Binance[F[_]: Monad: Clock](client: BinanceClient[F]) extends Exchan
         .map(_.filter(b => showEmpty || b.totalBalance > 0))
     }
 
-  def tradesHistory(pair: Option[TokenPair], paging: Paging): F[List[Trade]] = ???
+  def tradesHistory(pair: Option[TokenPair], paging: Paging): F[List[Trade]] = for {
+    params        <- BinanceTradeHistoryParams[F](pair, paging)
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala:29]: Explicit type annotation is redundant here

File: binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala [hunk 6]
```diff
@@ -24,14 +25,31 @@ final class Binance[F[_]: Monad: Clock](client: BinanceClient[F]) extends Exchan
...
+    tradesHistory <- client.tradesHistory(params.orNull).flatMap(_.traverse(_.make[F, Trade]))
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala:30]: Are you seriously passing `null` to Scala method? Please make `client.tradesHistory` accepting optional value.

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+object BinanceTradeHistoryParams {
+
+  def apply[F[_]: Sync](pairOp: Option[TokenPair], paging: Paging): F[Option[service.BinanceTradeHistoryParams]] =
+    for {
+      curPair <- pairOp.traverse(_.make[F, CurrencyPair])
+      params  <- Sync[F].delay(curPair.map(new service.BinanceTradeHistoryParams(_)))
+      _       <- Sync[F].delay(params.map(_.setStartTime(paging.startTime.map(l => new Date(l)).orNull)))
+      _       <- Sync[F].delay(params.map(_.setEndTime(paging.endTime.map(l => new Date(l)).orNull)))
+      _       <- Sync[F].delay(params.map(_.setLimit(paging.limit.map(int2Integer).orNull)))
+    } yield params
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala:24]:
```suggestion
  def apply[F[_]: Sync](tokenPair: TokenPair, paging: Paging): F[service.BinanceTradeHistoryParams] =
    for {
      curPair <- tokenPair.make[F, CurrencyPair]
      params = new service.BinanceTradeHistoryParams(curPair)
      _ <- Sync[F].delay(params.setStartTime(paging.startTime.map(l => new Date(l)).orNull))
      _ <- Sync[F].delay(params.setEndTime(paging.endTime.map(l => new Date(l)).orNull))
      _ <- Sync[F].delay(params.setLimit(paging.limit.map(int2Integer).orNull))
    } yield params
```

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import com.circlegain.exmods.core.make.syntax._
+import com.circlegain.exmods.core.models.Paging
+import com.circlegain.exmods.core.models.domain.TokenPair
+import org.knowm.xchange.binance.service
+import org.knowm.xchange.currency.CurrencyPair
+import tofu.syntax.monadic._
+
+object BinanceTradeHistoryParams {
+
+  def apply[F[_]: Sync](pairOp: Option[TokenPair], paging: Paging): F[Option[service.BinanceTradeHistoryParams]] =
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala:17]: You are passing optional argument to a method whose output completely depends on it. It's pointless, optionality of the `pairOp` arg should be processed at the call level.

**Review by oskin1 — CHANGES_REQUESTED** [review]

File: binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala (hunk 6 above)
Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala]: :confused:

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala (hunk 8 above)
Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/internal/models/BinanceTradeHistoryParams.scala]: :+1:

File: binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala [hunk 9]
```diff
@@ -24,14 +25,31 @@ final class Binance[F[_]: Monad: Clock](client: BinanceClient[F]) extends Exchan
...
+    for {
+      depositHistory  <- depositHistory(token, paging)
+      withdrawHistory <- withdrawHistory(token, paging)
+    } yield depositHistory ++ withdrawHistory
 }
 
 object Binance {
 
-  def make[I[+_]: Sync, F[_]: Sync: Clock](
+  def make[I[+_]: Sync, F[+_]: Sync: Clock](
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala:52]: Why declaring F as covariant?

Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/Binance.scala]: it's used to work around inference of OrderType

File: binance/src/main/scala/com/circlegain/exmods/binance/internal/instances.scala [hunk 10]
```diff
@@ -24,4 +28,72 @@ object instances {
...
+      case dto.Order.OrderType.ASK      => (Ask: OrderType).pure[F]
+      case dto.Order.OrderType.EXIT_ASK => (Ask: OrderType).pure[F]
+      case dto.Order.OrderType.EXIT_BID => (Bid: OrderType).pure[F]
+    }
+
+  implicit def makeToken[F[_]: Applicative]: Make[Currency, Token, F] =
+    Make(currency => Token.fromString(currency.getCurrencyCode).pure[F])
+
+  implicit def makeTrade[F[_]: Monad: Clock](implicit
+    orderTypeM: Make[dto.Order.OrderType, OrderType, F],
```
Comment (oskin1) [inline binance/src/main/scala/com/circlegain/exmods/binance/internal/instances.scala:47]: Why requiring these implicit makes?
You can just auto summon them in place using syntax.

Comment (eanea) [reply binance/src/main/scala/com/circlegain/exmods/binance/internal/instances.scala]: :+1:

**Review by oskin1 — APPROVED** [review]

**Pull request merged**